src/fabric/annealing.rs

```diff
@@ -94,56 +172,134 @@ impl MeasurePair {
                     return None;
                 }
-                let length = joint.location.distance(other_joint.location);
+                let length = self.joints[joint_index].location.distance(other_joint.location);
                 if length > length_limit {
                     return None;
                 }
-                Some(MeasurePair {
-                    alpha_index: joint.index,
+                Some(Pair {
+                    alpha_index: joint_index,
                     omega_index: other_joint.index,
-                    length,
+                    link: Link::Measure { length },
                 })
             })
             .map(|pair| (pair.key(), pair));
         self.pairs.extend(new_pairs);
     }
+
+    fn bow_tie_pulls(mut self) -> impl Iterator<Item=Pair> {
+        for interval in self.intervals.values() {
+            if interval.role != Role::Push {
+                continue;
+            }
+            let mut meeting_pairs: Vec<(Path, Path)> = vec![];
+            for alpha_path in self.paths_for(interval.alpha_index, 2) {
+                for omega_path in self.paths_for(interval.omega_index, 2) {
+                    if alpha_path.last() == omega_path.last() { // second one is the bridge
+                        meeting_pairs.push((alpha_path.clone(), omega_path.clone()))
+                    }
+                }
+            }
+            let [(alpha1, omega1), (alpha2, omega2)] = meeting_pairs.as_slice() else {
+                continue;
+            };
+            let diagonals = [
+                (alpha1.last_joint(), omega2.last_joint()),
+                (alpha2.last_joint(), omega1.last_joint())
+            ];
+            let candidates: Vec<(usize, usize)> = diagonals
+                .iter()
+                .filter_map(|&(a, b)| {
+                    match (self.joints[a].across_push(), self.joints[b].across_push()) {
+                        (Some(joint_a), Some(joint_b)) => {
+                            (!self.interval_exists(joint_a, joint_b)).then_some((a, b))
+                        }
+                        _ => None
+                    }
+                })
+                .collect();
+            if let &[(alpha_index, omega_index)] = candidates.as_slice() {
+                let link = Link::Pull {ideal: interval.ideal() / 3.0};
+                let pair = Pair { alpha_index, omega_index, link};
+                self.pairs.insert(pair.key(), pair);
+            }
+        }
+        self.pairs.into_values()
+    }
+
+    fn interval_exists(&self, a: usize, b: usize) -> bool {
+        if a < b {
+            self.intervals.contains_key(&(a, b))
+        } else {
+            self.intervals.contains_key(&(b, a))
+        }
+    }
+
+    fn get_interval(&self, a: usize, b: usize) -> Option<&Interval> {
+        if a < b {
+            self.intervals.get(&(a, b))
+        } else {
+            self.intervals.get(&(b, a))
+        }
+    }
```

Comment:
This redundant comparison makes me think you should key the intervals by `[usize; 2]` instead of `(usize, usize)` so that you can sort them before using them as a key.

Reply:
but then you gots to have your muterinos